core/src/main/scala/org/http4s/headers/Forwarded.scala

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.headers
+
+import java.net.{Inet4Address, Inet6Address}
+
+import cats.data.NonEmptyList
+import cats.syntax.either._
+import org.http4s._
+import org.http4s.util.{Renderable, Writer}
+
+object Forwarded
+    extends HeaderKey.Internal[Forwarded]
+    with HeaderKey.Recurring
+    with ForwardedRenderers
+    with parser.ForwardedModelParsing {
+
+  final case class Node(nodeName: Node.Name, nodePort: Option[Node.Port] = None)
+
+  protected[http4s] val NodeNameIpv4 = Node.Name.Ipv4
+
+  object Node {
+    def apply(nodeName: Name, nodePort: Port): Node = apply(nodeName, Some(nodePort))
+
+    sealed trait Name { _: Product => }
+
+    object Name {
+      case class Ipv4(address: Uri.Ipv4Address) extends Name
+      case class Ipv6(address: Uri.Ipv6Address) extends Name
+      case object Unknown extends Name
+
+      def apply(address: Uri.Ipv4Address): Name = Name.Ipv4(address)
+      def apply(address: Inet4Address): Name = apply(Uri.Ipv4Address.fromInet4Address(address))
+      def apply(a: Byte, b: Byte, c: Byte, d: Byte): Name = apply(Uri.Ipv4Address(a, b, c, d))
+
+      def apply(address: Uri.Ipv6Address): Name = Name.Ipv6(address)
+      def apply(address: Inet6Address): Name = apply(Uri.Ipv6Address.fromInet6Address(address))
+      def apply(a: Short, b: Short, c: Short, d: Short, e: Short, f: Short, g: Short, h: Short)
+          : Name = apply(Uri.Ipv6Address(a, b, c, d, e, f, g, h))
```

Comment:
A little late on this, but would we be better off with `from*` constructors instead of overriding `apply`?  I've soured a bit on overrides.  It's a fuzzy case: we usually `apply` if it's total, and `from*` if it's a `ParseResult`.  But we usually don't have so many options to make it total.

I could happily go either way here.

Reply:
Just thoughts... 

First, if we prefer to reduce (or completely get rid of) overrides here, we can safely drop these two constructors:
```
def apply(address: Uri.Ipv4Address): Name
def apply(address: Uri.Ipv6Address): Name
```
Because these two can be replaced with direct calls to `Name.Ipv4` and `Name.Ipv6` – these  `apply`-s are just shortcuts and thus there's no need to add extra `from*` methods for them.

Second, I'd suggest the following names for the remaining constructors:
```
def fromInet4Address(address: Inet4Address) // reflects `Uri.Ipv4Address.fromInet4Address`
def fromIpv4Address(a: Byte, b: Byte, c: Byte, d: Byte) // reflects `Uri.Ipv4Address.apply`

def fromInet6Address(address: Inet6Address): Name // reflects `Uri.Ipv6Address.fromInet6Address`
def fromIpv6Address(a: Short, ..., h: Short): Name // reflects `Uri.Ipv6Address.apply`
```
If you don't mind, I'm up to rename them.

Third, we could consider using different names like `of*` for creating totals when we'd like to avoid `apply` overrides. Just to disambiguate them from `from*`-s, which usually return `ParseResult`. Not sure if it will look consistently across the project, though.

Reply:
:+1: to all of this.  Yeah, I think I like the `of` convention.